Implement out-of-process SetThreadContext support for debugger detach

tommcdon opened the pull request:
This change handles a race between debugger detach and `Debugger::SendSetThreadContextNeeded` notifications. When the debugger calls `ICorDebugController::Detach`, we either delete or mark in-flight controllers for deletion which allows for further processing of in-flight breakpoint/single-step exceptions. If out-of-process SetThreadContext is needed is enabled, we could cause the debuggee to crash with an unhandled breakpoint exception while sending SetThreadContextNeeded flares after we detached. This change causes the detach operation to track SetThreadContextNeeded operations before signaling completion. Note the change only takes affect if CET debugging is enabled, and it only tracks notification count and state if we are in the process of detaching. This change handles early process exit during attach as well.  
Fixes #116870

tommcdon committed:
Implement out-of-process detach support

tommcdon committed:
Fix build break

tommcdon committed:
Handle early process termination during detach in out-of-process debugging

thaystg reviewed (approved):
LGTM

tommcdon committed:
Track active controller ref count for detach through DebuggerController creation and desctruction and DispatchPatchOrSingleStep

tommcdon commented on src/coreclr/debug/ee/controller.h:
```diff
@@ -1157,6 +1157,19 @@ class DebuggerController
+    static int IncrementActiveDispatchedExceptions() { LIMITED_METHOD_CONTRACT;  return (int)InterlockedIncrement(&g_cActiveDispatchedExceptions); }
+    static int DecrementActiveDispatchedExceptions() { LIMITED_METHOD_CONTRACT;  return (int)InterlockedDecrement(&g_cActiveDispatchedExceptions); }
+    static int GetNumPendingControllers();
+    static int GetDispatchedFlares() {LIMITED_METHOD_DAC_CONTRACT;  return g_cDispatchedFlares; }
```
> For example imagine a thread that was doing a patch skip has already exited DPOSS, deleted the DebuggerPatchSkip controller and it is just about to send the flare in FirstChanceNativeException.
> there is also a race between when the debugger starts counting flares received and when the debuggee computes the count to send in the DetachComplete event.

We are expecting the runtime to be synchronized before we start the detach operation. We also take the controller lock to counting the controllers - which gives us some guarantees we have here that help us to get an accurate count. There are 3 checks inside of DPOSS to account for a thread that entered without detach being active and resuming from the controller lock with detach started. In my testing I didn't find any issues - I spun up threads hitting breakpoints in a while (true) loop and then performed a detach. I observed ~6000 threads in various places and no observed no ref count issues with flares. I was able to detach/re-attach over and over with no errors. With that said I agree there is conceptually a small time window between synchronization and threads escaping from DPOSS and sending flares. 

>I suspect if we want to resolve these race conditions we need to make the debugger side maintain a count on its own.

I will make changes to enable ref counting flares continuously rather than only during detach. This should eliminate concerns with race conditions. I will also move the ActiveDispatchedExceptions counting logic up higher on the stack to further reduce concerns with timing.

noahfalk replied:
> We are expecting the runtime to be synchronized before we start the detach operation

I don't believe that synchronizing the runtime guarantees anything about the state of threads hitting our breakpoints. As far as I know its legal to set breakpoints in pre-emptive code which means some of those threads might still be running in parallel. If we can prove that breakpoints never get set in pre-emptive code that would change things but the comments don't suggest that is the case:
https://github.com/dotnet/runtime/blob/main/src/coreclr/debug/ee/controller.cpp#L4412-L4415

tommcdon replied:
The updated change I just pushed continuously tracks the number of flares and threads in DC:DNE. It removes the tracking of the number of outstanding controllers, and replaces it with a simple check on `g_controllers`.  This is the same test case I used before - just sharing what I have been using to validate the fix - to test, I put a breakpoint on the `if (!doExit...` line, stepping several times, detach, re-attach and repeat.  Do so, I have observed 6000+ threads on my machine with approx. 800 threads queued up under various places.  I also set doExit to true in the debugger, and then either continue or detach to test the program exit scenario. I also tested a one-line helloworld app, with stopping on the Console.WriteLine, then detaching (to test early exit). I also tested both with and without out of process setthreadcontext enabled.  Please let me know if this does not address the concerns or if there are any other issues - happy to address them!
```csharp
using System;
using System.Threading;

class Program
{
	static void Main()
	{
		Console.WriteLine("waiting for debugger to attach...");
		while (!System.Diagnostics.Debugger.IsAttached)
		{
			// Wait for debugger to attach
			Thread.Sleep(100); // 100ms
        }
        Console.WriteLine("Starting thread spawner...");
		bool doExit = false;
		int i = 0;
        while (!doExit)
		{
			Thread t = new Thread(() =>
			{
				// Thread exits quickly
				// Simulate quick work
				Thread.Sleep(1); // 1ms
				if (!doExit && !System.Diagnostics.Debugger.IsAttached)
				{
					//Environment.Exit(0);
                    //doExit = true;
                }
            });
			t.Start();
			// Optionally throttle thread creation to avoid overwhelming the system
			if (i++ > 1000)
				Thread.Sleep(1); // 1ms between spawns
		}
		Console.WriteLine("Debugger detached");
	}
}
```

noahfalk replied:
Unforetunately I think you resolved some of the issues but races are still present. Sorry I wasn't clear before :( Let me take another stab at describing some of the issues.

### Controllers can be created after the code confirms there aren't any

[Thread 1] currently just received the DB_IPCE_DETACH_FROM_PROCESS event and it is about to call HandleEvent().
[Thread 2] is a newly spawned in thread that has just called Debugger::ThreadCreated(), then the OS switches it out for a bit.
[Thread 1] resumes and runs CanSendDetach(). It goes inside the ControllerLock at [line 4409](https://github.com/dotnet/runtime/pull/118849/files#diff-56c927cca69b678fd1fdbeab66e14e38a36fc6f913cb3d30af7ac37859ac2d3fR4393), and it has observed there are no controllers nor exceptions. Now CanSendDetach() exits the lock and returns to its caller Debugger::HandleDebuggerEvent() where the OS switches it out for a while.
[Thread 2] resumes running Debugger::ThreadCreated() which invokes new DebuggerThreadStarter(), which calls Controller::Controller() that links a new controller.
[Thread 1] resumes and sends the DB_IPCE_DETACH_FROM_PROCESS_RESULT. Even though there were no controllers when thread 1 checked earlier there are controllers now. The consequences of this particular case don't look that bad if it continues - we leak the controller.

### The dispatch count can increase after we've sent it.

Here is a worse example that looks like it could crash

[Thread 1] currently received the DB_IPCE_DETACH_FROM_PROCESS event and it is about to call HandleEvent().
[Thread 2] is executing in preemptive mode, hit some breakpoint opcode and has just reached the start of Debugger::DispatchNativeException to notify the debugger.
[Thread 3] is inside a debugger patch skip buffer. It has just exited from the call to Debugger::DispatchNativeException(). The DispatchFlareCount was incremented to 1 but the flare hasn't yet been sent.
[Thread 1] runs DebuggerController::DeleteAllControllers() which removes all controllers and patches. CanSendDetach() returns true and SendDetachComplete() sends the DB_IPCE_DETACH_FROM_PROCESS_RESULT with flare count = 1
[Thread 2] continues executing Debugger::DispatchNativeException. Inside of ScanForTriggers it observes [the address is no longer patched](https://github.com/dotnet/runtime/blob/a6740f49b59afca0dd1d1b2b6c8bdd01ece1fb88/src/coreclr/debug/ee/controller.cpp#L2757-L2762) so it returns DPOSS_USED_WITH_NO_EVENT which sets fDebuggers=TRUE back in [DispatchNativeException](https://github.com/dotnet/runtime/pull/118849/files#diff-56c927cca69b678fd1fdbeab66e14e38a36fc6f913cb3d30af7ac37859ac2d3fR4665) which increments the flare count to 2. The thread continues out of DispatchNativeException() and sends the flare.
[Thread 3] proceeds to send its own flare but the debugger is already detached because it was only waiting for 1 flare. Failing to SetThreadContext it continues executing in the patch buffer and crashes.

### The debugger can observe flares before the debuggee resets the count

This one looks like the debugger would wait at detach for flares that are never coming.

Code can call into DispatchNativeException prior to the debugger sending the DB_IPCE_ATTACHING event. If something triggered a 1st chance breakpoint (maybe code in some native library?) that will run through DispatchExceptionHook, observe that g_patchTableValid = false, which sets fDispatch=false, which sets fDebuggers=TRUE. The flare will get sent to the debugger incrementing its count to 1 and its a race between the ResetDispatchFlares() and IncrementDispatchedFlares() to determine whether the debuggee includes the flare in its own count.


While I do think its possible to keep finding and fixing these issues, having the debuggee doing the counting seems complex and fragile. It takes a bunch of time to reason about all the multi-threaded execution orderings that could occur as well as when DispatchNativeException will send flares. Subtle changes in the debuggee code could re-introduce new issues even if we resolved every current one (the list above probably isn't exhaustive, I just stopped looking after finding a few). My suggestion is that we stop trying to create complex logic to track flare counts and instead handle this almost entirely on the debugger side. I suspect if we do something like this we can end up with a simpler and more robust solution:
1. On the debuggee side, include a flag in the SetThreadContext flare that indicates the thread is being moved into the patch buffer.
2. Once the IPCE_DETACH_FROM_PROCESS response is received, the debugger should start ignoring any SetContext flares that have that flag set. If the debuggee sees a SetContext into the patch buffer was ignored (execution continues after the flare) it should set the IP back to the original instruction, clear the SS flag, and let the VEH continue. During detach we already deleted all the patches so the original address shouldn't have a breakpoint now.
3. In the debugger, don't win32 detach until every thread either has never sent a flare, or it sent a flare without that flag set. In other words, no threads are currently in the patch buffer.
4. All the other debuggee changes go away - no controller counts, exception counts, flare counts, or detach safety checks. From the debuggee side managed debugger detach is the same as before except now the SetThreadContextNeededFlare keeps working for a while after managed detach is already done.

tommcdon committed:
Continuously track active dispatched exceptions and flares sent

tommcdon committed:
Out-of-process detach support for CET

Debugger events are tracked out of process to determine which threads have outstanding SetThreadContextNeeded flares

tommcdon committed:
Revert "Continuously track active dispatched exceptions and flares sent"

This reverts commit dd6d28cd60bd73097116ace4c7dc72658aeb796b.

tommcdon committed:
Revert "Track active controller ref count for detach through DebuggerController creation and desctruction and DispatchPatchOrSingleStep"

This reverts commit 367c63847ed3f0ce49639a4e60d6e7ca4eb44687.

tommcdon committed:
Revert "Fix build break"

This reverts commit 240b37210562d629c2997913dad2e8b40884f6e5.

tommcdon committed:
Revert "Implement out-of-process detach support"

This reverts commit d80b6c4475a93cf1a1257087eddf379e62a0be6e.

tommcdon committed:
Remove unused enum

tommcdon committed:
Break out of detach if  timed out or process exited

tommcdon committed:
Merge branch 'main' into dev/tommcdon/fix_cet_detach

noahfalk reviewed (approved):
Yeah, this is getting much easier to reason about. I put a few suggestions inline but overall looks good!

tommcdon commented on src/coreclr/debug/di/process.cpp:
```diff
@@ -11915,6 +12034,13 @@ void CordbWin32EventThread::Win32EventLoop()
             hrShim = m_pShim->HandleWin32DebugEvent(&event);
         }
+#ifdef OUT_OF_PROCESS_SETTHREADCONTEXT
+        if (fIsDetaching)
```
I propose that we pulse the detach loop inside of HandleSetThreadContextNeeded and HandleInPlaceSingleStep, and keep the existing mechanism poll mechanism from CordbProcess::DetachShim. The reason is that once we drain all SetThreadContextNeeded/HandleInPlaceSingleStep events, it should be safe to detach, but in the event we don't we have the 5 second timeout.  If we start the 5 second timer immediately and wait for the m_detachIsSafeEvent event, it may be possible that we take longer than 5 seconds to detach while we are still receiving valid SetThreadContextNeeded/HandleInPlaceSingleStep callbacks.

tommcdon committed:
Add ClearSetIP to SetThreadContextNeeded event,
Refactor HandleSetThreadContextNeeded for improved readability

tommcdon committed:
Handle case when out of process setthreadcontext is oped-out but available on the platform

noahfalk commented on src/coreclr/debug/di/rspriv.h:
```diff
@@ -2920,10 +2920,14 @@ struct DbgAssertAppDomainDeletedData
+    HANDLE m_hThread = INVALID_HANDLE_VALUE; // Handle to the unmanaged thread, used for suspending and resuming the thread
+    CORDB_ADDRESS_TYPE *m_pPatchSkipAddress = NULL; // If non-NULL, this is the address to which we should set the IP when resuming the thread.
+    DWORD m_dwSuspendCount = 0; // The suspend count of the thread when we last checked it.
+    bool m_pendingSetIP = false; // Set to true if there is a breakpoint or single-step operation on target thread that doesn't have a vmTHread
```
```suggestion
    bool m_pendingSetIP = false; // Set to true if there is a breakpoint or single-step operation on target thread
```
That vmThread part isn't accurate is it?

tommcdon committed:
PR feedback on code comment

tommcdon commented:
/backport to release/10.0
